sky/serve/serve_utils.py

```diff
@@ -41,6 +42,42 @@
     f' to check all valid replica id.{colorama.Style.RESET_ALL}')
 
 
+class ThreadSafeDict(dict):
+    """A thread-safe dict."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._lock = threading.Lock()
+
+    def __getitem__(self, __key: Any) -> Any:
+        with self._lock:
+            return super().__getitem__(__key)
+
+    def __setitem__(self, __key: Any, __value: Any) -> None:
+        with self._lock:
+            return super().__setitem__(__key, __value)
+
+    def __delitem__(self, __key: Any) -> None:
+        with self._lock:
+            return super().__delitem__(__key)
+
+    def __len__(self) -> int:
+        with self._lock:
+            return super().__len__()
+
+    def __contains__(self, __key: object) -> bool:
+        with self._lock:
+            return super().__contains__(__key)
+
+    def items(self):
+        with self._lock:
+            return super().items()
+
+    def values(self):
+        with self._lock:
+            return super().values()
+
+
```

Comment:
it might be safer to not directly inherit the built-in python dict but create a new class containing a python dict and only add some methods (e.g., `items()`) used in our code.

otherwise, in the future if we use other built-in methods of `Dict` that are missing here are still not thread safe?

Reply:
for example:

```
class ThreadSafeDict(object):
    def __init__(self):
        self._dict = {}
        self._lock = threading.Lock()
    
    def set(self, key, value):
        with self._lock:
            self._dict[key] = value
    
    def get(self, key):
        with self._lock:
            return self._dict.get(key)
    
    def delete(self, key):
        with self._lock:
            if key in self._dict:
                del self._dict[key]
    
    def contains(self, key):
        with self._lock:
            return key in self._dict

    def items(self):
        with self._lock:
            return self._dict.items()
    
```

Reply:
Make sense... This way it will report an function not defined error

Reply:
Done. Thanks!